### PR TITLE
Support node taints configuration

### PR DESCRIPTION
**Pull request by orangedeng**
**Problem:**
We can not set node taints in RKE node config.

**Solution:**
Sync taints from config in `SyncLabelsAndTaints` function

Related issue:
https://github.com/rancher/rancher/issues/13972

### CODE REVIEW

**Comment by gitlawr** [comment]
Can we do it in a similar way as labels - Prepare `ToAdd/DelLabels` early before the `SyncLabelsAndTaints` function.

File: cluster/cluster.go [hunk 1]
```diff
@@ -84,6 +84,9 @@ const (
 	WorkerThreads = util.WorkerThreads
 
 	serviceAccountTokenFileParam = "service-account-key-file"
+
+	unschedulableEtcdTaintKey    = "node-role.kubernetes.io/etcd=:NoExecute"
+	unschedulableControlTaintKey = "node-role.kubernetes.io/controlplane=:NoSchedule"
```
Comment (gitlawr) [inline cluster/cluster.go]: Can you show reference for these reserved taints

Comment (orangedeng) [reply cluster/cluster.go]: I put the similar taints together. I think we will understand better in this way.

**Comment by orangedeng** [comment]
Move the logic to `ReconcileCluster`, put them in function `syncTaints`. Updated.

File: cluster/reconcile.go [hunk 2]
```diff
@@ -377,3 +382,74 @@ func isEtcdPlaneReplaced(ctx context.Context, currentCluster, kubeCluster *Clust
...
+	}
+	return rtn
+}
+
+func getTaintKey(taint v3.RKETaint) string {
+	return fmt.Sprintf("%s=:%s", taint.Key, taint.Effect)
+}
+
+func getTaintValue(taint v3.RKETaint) string {
+	return fmt.Sprintf("%s=%s:%s", taint.Key, taint.Value, taint.Effect)
```
Comment (gitlawr) [inline cluster/reconcile.go]: Can we use `taint.Value` directly?

Comment (orangedeng) [reply cluster/reconcile.go]: I am adding this value to `toAddTaints` and `toDelTaints` array.

**Comment by gitlawr** [comment]
LGTM

**Comment by cjellick** [comment]
@galal-hussein can you review?

File: vendor/github.com/rancher/types/apis/management.cattle.io/v3/rke_types.go [hunk 3]
```diff
@@ -174,7 +174,7 @@ type RKEConfigNode struct {
 	// Node Labels
 	Labels map[string]string `yaml:"labels" json:"labels,omitempty"`
 	// Node Taints
-	Taints []v1.Taint `yaml:"taints" json:"taints,omitempty"`
+	Taints []RKETaint `yaml:"taints" json:"taints,omitempty"`
```
Comment (moelsayed) [inline vendor/github.com/rancher/types/apis/management.cattle.io/v3/rke_types.go]: Why are you using a custom type ?

**Comment by moelsayed** [comment]
@orangedeng We are doing this(or something very similar) near the end of provisioning the cluster https://github.com/rancher/rke/blob/master/cmd/up.go#L176. It's not a good idea to do it at the reconcile stage since it will affect nodes the might be added/deleted/modified later. 

Also, there are several edge cases around this process, and it's kind of a bottleneck when you have many nodes. Would it be possible to incorporate your changes in the current logic or extract the taints related code to avoid doing the work twice ?

**Comment by orangedeng** [comment]
Updated

**Comment by cjellick** [comment]
@moelsayed please re-review

**Comment by deniseschannon** [comment]
https://github.com/rancher/rke/issues/157

**Comment by cjellick** [comment]
worth having another set of eyes on this. @superseb please review

**Comment by cjellick** [comment]
@superseb please review

**Comment by daxmc99** [comment]
Does this support node taints on `rke up` with a config file?

**Comment by orangedeng** [comment]
Yes.

**Comment by orangedeng** [comment]
Rebased on master

**Pull request merged**